acceptance/api_v1_services_mutation_test.go

```diff
@@ -0,0 +1,787 @@
+				cleanupService(service)
+			})
+
+			It("returns a 'conflict' when the service already exists", func() {
+				response, err := Curl("POST",
+					fmt.Sprintf("%s/api/v1/orgs/%s/services",
+						serverURL, org),
+					strings.NewReader(fmt.Sprintf(`{
+					    "name": "%s",
+					    "class": "meh",
+					    "plan": "meh",
+					    "waitforprovision": true
+					}`, service)))
+				Expect(err).ToNot(HaveOccurred())
+				Expect(response).ToNot(BeNil())
+
+				defer response.Body.Close()
+				bodyBytes, err := ioutil.ReadAll(response.Body)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(response.StatusCode).To(Equal(http.StatusConflict), string(bodyBytes))
+				Expect(string(bodyBytes)).To(Equal("Service '" + service + "' already exists\n"))
+			})
+		})
+
+		It("returns a 'not found' when the class does not exist", func() {
+			response, err := Curl("POST",
+				fmt.Sprintf("%s/api/v1/orgs/%s/services",
+					serverURL, org),
+				strings.NewReader(fmt.Sprintf(`{
+				    "name": "%s",
+				    "class": "meh",
+				    "plan": "meh",
+				    "waitforprovision": true
+				}`, service)))
+			Expect(err).ToNot(HaveOccurred())
+			Expect(response).ToNot(BeNil())
+
+			defer response.Body.Close()
+			bodyBytes, err := ioutil.ReadAll(response.Body)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(response.StatusCode).To(Equal(http.StatusNotFound), string(bodyBytes))
+			Expect(string(bodyBytes)).To(Equal("Service class 'meh' does not exist\n"))
+		})
+
+		It("returns a 'not found' when the plan does not exist", func() {
+			response, err := Curl("POST",
+				fmt.Sprintf("%s/api/v1/orgs/%s/services",
+					serverURL, org),
+				strings.NewReader(fmt.Sprintf(`{
+				    "name": "%s",
+				    "class": "mariadb",
+				    "plan": "meh",
+				    "waitforprovision": true
+				}`, service)))
+			Expect(err).ToNot(HaveOccurred())
+			Expect(response).ToNot(BeNil())
+
+			defer response.Body.Close()
+			bodyBytes, err := ioutil.ReadAll(response.Body)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(response.StatusCode).To(Equal(http.StatusNotFound), string(bodyBytes))
+			Expect(string(bodyBytes)).To(Equal("Service plan 'meh' does not exist for class 'mariadb'\n"))
+		})
+
+		Context("Creation", func() {
+			AfterEach(func() {
+				cleanupService(service)
+			})
+
+			It("creates the catalog service and waits for it to be provisioned", func() {
+				response, err := Curl("POST",
+					fmt.Sprintf("%s/api/v1/orgs/%s/services",
+						serverURL, org),
+					strings.NewReader(fmt.Sprintf(`{
+					    "name": "%s",
+					    "class": "mariadb",
+					    "plan": "10-3-22",
+					    "waitforprovision": true
+					}`, service)))
+				Expect(err).ToNot(HaveOccurred())
+				Expect(response).ToNot(BeNil())
+
+				defer response.Body.Close()
+				bodyBytes, err := ioutil.ReadAll(response.Body)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(response.StatusCode).To(Equal(http.StatusCreated), string(bodyBytes))
+				Expect(string(bodyBytes)).To(Equal(""))
```

Comment:
the expectations are similar to the test below. Maybe we should check that the service is there indeed? On the other hand I can't think of a simple way to make this kind of test reliable, it will be "racy" unless we find some way to control (hold and release) the provisioning process on the server.

Reply:
I should add checks for service presence to both indeed. Should be able to take that from the cli service creation tests.

Reply:
ok try that but be careful of race conditions. Eventually the service will be there and depending on how long that takes to happen, you may get different results when you check.

Reply:
I am not sure where you see the race. This test has `waitforprovision: true` in the request. The curl will literally not return before the service exists, because the server waits for it to exist. The next test has it `wfp:false`, so the server returns immediately, and thus the test has to wait for the service to exist (using `Eventually`) before it can continue.

Reply:
According to the test description ("creates the catalog service and waits for it to be provisioned"), this test also checks that waiting is indeed happening. Let's say we change the server code in the future and somehow we break the part that waits for the service to be created and instead it always returns immediately. This test wouldn't fail in that case. Even if you tested that the service was created in the test, you won't know if this happened before or after your request was completed.